newIDE/app/src/CommandPalette/CommandsScopedContext.js

```diff
@@ -9,31 +13,72 @@ type Props = {
   active: boolean,
 };
 
+class ScopedCommandManager implements CommandManagerInterface {
+  commands: { [string]: Command };
+  _centralManager: CommandManager;
+  isActive: boolean;
+
+  constructor(centralCommandManager: CommandManager) {
+    this.commands = {};
+    this.isActive = false;
+    this._centralManager = centralCommandManager;
+  }
+
+  setActive = (active: boolean) => {
+    this.isActive = active;
+  };
+
+  registerCommand = (commandName: string, command: Command) => {
+    this.commands[commandName] = command;
+    if (this.isActive)
+      this._centralManager.registerCommand(commandName, command);
+  };
+
+  deregisterCommand = (commandName: string) => {
+    delete this.commands[commandName];
+    if (this.isActive) this._centralManager.deregisterCommand(commandName);
+  };
+
+  registerAllCommandsToCentralManager = () => {
+    Object.keys(this.commands).forEach(commandName => {
+      this._centralManager.registerCommand(
+        commandName,
+        this.commands[commandName]
+      );
+    });
+  };
+
+  deregisterAllCommandsFromCentralManager = () => {
+    Object.keys(this.commands).forEach(commandName => {
+      this._centralManager.deregisterCommand(commandName);
+    });
+  };
+
+  getAllNamedCommands = () => {
+    return Object.keys(this.commands).map<NamedCommand>(commandName => {
+      const cmd = this.commands[commandName];
+      return { ...cmd, name: commandName };
+    });
+  };
+}
+
 const CommandsContextScopedProvider = (props: Props) => {
   const centralManager = React.useContext(CommandsContext);
-  const scopedManager = useRefInit(() => new CommandManager(true));
+  const scopedManager = useRefInit(
+    () => new ScopedCommandManager(centralManager)
+  );
```

Comment:
This part gives a Flow error. Since CommandsContext is now typed as containing a `CommandsManagerInterface`, Flow does not know if `centralManager` here is a `CommandManager` or a `ScopedCommandManager`, while the constructor of `ScopedCommandManager` strictly takes a `CommandManager`.

Should I change the type of `ScopedCommandManager._centralManager` as a `CommandManagerInterface`? It kinda hints towards layered facades (like more tabs inside a tab), although that's probably not very useful.

Reply:
Yes I think the type should be a `CommandManagerInterface`. The "beauty" of this is that as you mentioned, you can layer the facades - so that you could use a `CommandsContextScopedProvider` inside a `CommandsContextScopedProvider` and it would work perfectly fine.

Not very useful for now, but possible in theory and might get some usage in the future!